### DISCUSSION TITLE
Boris - A RustAnalyzer based Ownership and Borrowing Visualizer

### DISCUSSION THREAD

**Discussion by ChristianSchott** [rust-lang/rust-analyzer#16437] [Show and tell]
Hello
I am not sure if this is the right place to post this, but I thought this may be interesting for the maintainers here..

As part of my master's thesis, I am currently developing a tool for generating interactive visualizations of ownership and borrowing utilizing the RA crate.

![app_example](https://github.com/rust-lang/rust-analyzer/assets/56631269/651143b6-76b6-44ba-9fc2-780f0e5738a9)

https://github.com/ChristianSchott/boris

It is basically a custom move/ reference tracking analysis based on RA's 'new' MIR lowering support (@HKalbasi). 
All of this is of course still very alpha, but I thought you might want to see something based on your project :)

As this is part of a research project, there is also an ongoing survey to evaluate the program https://opnform.com/forms/visualizing-ownership-and-borrowing-in-rust-programs-nseo4z
It only has 7 short questions and contains the interactive visualizations embedded in the survey.. so if you want to provide some feedback, that would really help me :)

Best regards
Christian

**Comment by Scott169** [comment]
Great work, this would be useful for beginners to understand the concept of lifetime.
